Atomic.md

```diff
@@ -0,0 +1,270 @@
+
+    -   An abandoned folder is in .temp that can be safely
+        overwritten/deleted when the enumeration is retried
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Folder is marked as in virtualization root but not empty.
+
+        -   A partial placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   Folder is marked as empty but not in virtualization root
+
+        -   A partial placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   The rename() to copy the completed placeholder from .temp into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+WritePlaceholderFile:
+
+-   Fopen() fails to create a file
+
+    -   We can simply retry and have this succeed. Nothing was created
+        in .temp
+
+-   Ftruncate fails to expand the file to the correct size (or is never
+     reached because of a crash)
+
+    -   We can simply retry and have this succeed. The file written in
+        .temp will be overwritten/deleted and recreated
+
+-   Memcpy'ing the providerID xattr fails (or is never reached because
+     of a crash)
+
+    -   We have a partial placeholder file in .temp that will be
+        overwritten by a retry.
+
+-   Memcpy'ing the contentID xattr fails (or is never reached because of
+     a crash)
+
+    -   We have a partial placeholder file in .temp that will be
+        overwritten by a retry.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   We have a partially created placeholder written into .temp that
+        will be overwritten on retry
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Folder is marked as in virtualization root but not empty.
+
+        -   A partial placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+    -   Folder is marked as empty but not in virtualization root
+
+        -   A partial placeholder is in .temp that can be safely
+            overwritten/deleted when the enumeration is retried
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   We can retry and have this succeed as the placeholder hasn't
+        left .temp yet
+
+    -   The rename() to copy the completed placeholder from .temp into
+        the virtualization root fails
+
+        -   We can retry from this error since nothing ever appeared in
+            the virtualization root. The placeholder will be overwritten
+            as part of the retry
+
+All of the potential holes identified above are filled by this design,
+so I believe this to be our best bet to create placeholders in an atomic
+matter.
+
+Miscellanea/Things to Validate
```

Comment:
Something else to validate:  Do we need to ensure that that flags and xattrs are flushed to disk before renaming the placeholder?